server.go

```diff
@@ -282,12 +282,15 @@ func (s *Server) sendResponse(t transport.ServerTransport, stream *transport.Str
 }
 
 func (s *Server) processUnaryRPC(t transport.ServerTransport, stream *transport.Stream, srv *service, md *MethodDesc) (err error) {
+	ctx, cancel := context.WithCancel(stream.Context())
```

Comment:
*Stream.cancel is called from *http2Server.closeStream, which is called from a few different places, notably *http2Server.WriteStatus.  But WriteStatus returns early if the stream is already in state streamDone, which is set in 3 places.  I cannot tell whether those other code locations lead to calling *Stream.cancel, but it seems like they should.